blog/2021-06-17-activity-timeouts.md

```diff
@@ -0,0 +1,258 @@
+
+```java
+@WorkflowInterface
+public interface SimpleWorkflow {
+    @WorkflowMethod
+    String simpleWorkflowMethod(String someArg);
+}
+
+public static class SimpleWorkflowImpl implements SimpleWorkflow {
+    private final SimpleActivities activities =
+        Workflow.newActivityStub(
+            SimpleActivities.class,
+            ActivityOptions.newBuilder()
+              .setScheduleToCloseTimeout(Duration.ofSeconds(500))
+              .setStartToCloseTimeout(Duration.ofSeconds(60))
+              .setHeartbeatTimeout(Duration.ofSeconds(10))
+              .build());
+
+    @Override
+    public String simpleWorkflowMethod(String name) {
+      return activities.simpleActivity("Hello", name);
+    }
+}
+```
+
+</TabItem>
+</Tabs>
+
+Behind the scenes, the SDK transforms this to a `ScheduleActivity` Command, which is sent to the Temporal Server. This Command includes various metadata, including the activity type (`Foo`), activity task queue (`Bar`), activity ID, and `RetryPolicy` (if not specified, Temporal uses a default policy).
+
+### Step 2 - Temporal Server
+
+Receiving the Command, Temporal Server then sets up the mutable state for that workflow and activity ID, and also adds an `ActivityTask` to the `Bar` Activity Queue.
+There is an atomic guarantee that these both happen together, to prevent race conditions. 
+We explained why this is important and how Temporal accomplishes this in [Designing A Workflow Engine](https://docs.temporal.io/blog/workflow-engine-principles/).
+
+> The activity is now in `SCHEDULED` state.
+
+### Step 3 - Activity Worker 
+
+An Activity Worker that has been polling for the `Bar` activity queue picks up the `ActivityTask` and begins execution.
+
+> The activity is now in `STARTED` state.
+
+### Step 4 - Temporal Server
+
+Once the activity finishes successfully, the Activity Worker sends a `CompleteActivityTask` message (together with the result of the activity) to Temporal Server, which now gives control back to the Workflow Worker to continue to the next line of code and repeat the process.
+
+> The activity is now in `CLOSED` state.
+
+We have just described the "Happy Path" of an activity. However, what happens when a worker crashes midway through an execution?
+
+## StartToClose Timeout
+
+We use the `StartToCloseTimeout` to control the maximum amount of time **a single activity invocation** can take. We recommend **always** setting this timeout.
+
+![image](https://user-images.githubusercontent.com/6764957/122290108-fff64000-cf25-11eb-92b3-0533e41c3fee.png)
+
+The classical example for how `StartToClose` becomes relevant is when an activity has been picked up from the activity queue (`STARTED` state) but crashes midway (so it never reaches `CLOSED` state). 
+
+- Without a timeout configured, Temporal would never proactively timeout this activity to initiate a retry. The activity becomes "stuck" and the end user experiences an indefinite delay of their work with no feedback.
```

Comment:
An activity cannot be stuck if (ScheduleToClose timeout configured) or ((ScheduleToStart timeout and StartToClose timeout configured) or (ScheduleToStart timeout and Heartbeat timeout configured))